### ISSUE TITLE
[ENH]: Support diamond (45 deg rotated) scatterplots

### ISSUE THREAD

**Issue by ellisonbg** [matplotlib/matplotlib#24501]
### Problem

Traditional scatter plots lead of confusion between causation and correlation.

### Proposed solution

This is a nice paper that proposes rotating scatterplots by 45 degreess to help reduce confusion about causation and correlation. 

https://arxiv.org/abs/1809.09328

Would be nice to have this implemented in matplotlib as an option to scatter.

**Comment by tacaswell** [comment]
https://matplotlib.org/stable/gallery/axisartist/demo_floating_axes.html has an example that does this.

**Comment by tacaswell** [comment]
Making this easier seems like a good idea, but it should be on the `Axes` object, not a keyword argument to `scatter`.  We do have a few plotting commands that mess with their host Axes (looking at you `imshow`) but I think that those are generally not a good idea as it is crossing concerns too much.

**Comment by jklymak** [comment]
I would handle this as a different projection, like polar or stereographic.

**Comment by anntzer** [comment]
I think we should try to bring https://matplotlib.org/stable/gallery/axisartist/demo_floating_axes.html more in line with standard matplotlib semantics (and better expose it), as that should support the use case suggested here.  I've been slowly trying to work on that...

**Comment by jklymak** [comment]
I would advocate for the opposite approach and treat floating axes as a proof of concept and introduce a new feature into the main library.  That allows a clean API break, sort of like what we did for inset_axes.

**Comment by anntzer** [comment]
Sure, that's a good idea too, but I do believe that moving floating_axes towards standard mpl semantics will help if someone wants at some point to take whatever they can from that module and reuse it for a (partially) new implementation of the feature for the main lib.

**Comment by oscargus** [comment]
Also documenting `FloatingAxes` et al. would be a step in the right direction.

**Comment by oscargus** [comment]
I tried to create a diamond plot using the example pointed out above, but gets stuck at rotation the tick labels back to vertical (or rotating them at all). I think I've tried most methods, but to no avail. Any pointer to which approach I am missing?

**Comment by anntzer** [comment]
I don't think unrotated labels are supported as is by axisartist(?).  Probably a reasonable feature to add, they look reasonably nice in the linked paper (attn @leejjoon).

**Comment by leejjoon** [comment]
> I don't think unrotated labels are supported as is by axisartist(?). Probably a reasonable feature to add, they look reasonably nice in the linked paper (attn @leejjoon).

You are correct, unrotated labels are not supported. It should be straight forward to add the feature though. On the other, if we want to make it more general, we would need an algorithm that automatically adjust 'ha' and 'va' accordingly.

**Comment by oscargus** [comment]
This is what I got after some more fiddling:

![image](https://user-images.githubusercontent.com/8114497/204903305-5befce30-b7e1-497e-a29d-b130a2fef66d.png)

Not perfect, one should probably figure out the rotation mode and alignments a bit better.

Worthwhile creating an example?

Code:
```
from matplotlib.transforms import Affine2D
import mpl_toolkits.axisartist.floating_axes as floating_axes
import numpy as np
from mpl_toolkits.axisartist.grid_finder import MaxNLocator
import matplotlib.pyplot as plt

fig = plt.figure(figsize=(4, 4))
fig.subplots_adjust(wspace=0.3, left=0.05, right=0.95)

tr = Affine2D().rotate_deg(45)
grid_helper = floating_axes.GridHelperCurveLinear(
    tr, extremes=(0, 1, 0, 1),
    grid_locator1=MaxNLocator(4),
    grid_locator2=MaxNLocator(4))
ax1 = fig.add_subplot(
        111, axes_class=floating_axes.FloatingAxes, grid_helper=grid_helper)
ax1.grid()

rot_ax = ax1.get_aux_axes(tr)

rot_ax.scatter(np.random.rand(10, 1), np.random.rand(10, 1))
ax1.axis['bottom'].major_ticklabels.set(rotation=-45, ha='left', va='top')
ax1.set_xlabel('Variable A')
ax1.axis['bottom'].label.set(rotation=-45, ha='left', va='top')
ax1.axis['left'].major_ticklabels.set(rotation=45, ha='right', va='top')
ax1.set_ylabel('Variable B')
ax1.axis['left'].label.set(rotation=45, ha='right', va='top')
```
It wasn't obvious to me that `ax1` contained the labels, not `rot_ax`.

**Comment by anntzer** [comment]
> It wasn't obvious to me that ax1 contained the labels, not rot_ax.

As a side note, I have some WIP trying to flatten everything into rot_ax (offering an API that allows one to hide completely ax1 away), but it's still very much a WIP.

**Comment by anntzer** [comment]
At least an example seems reasonable to add.

**Comment by timhoffm** [comment]
Yes, add it to https://matplotlib.org/stable/gallery/index#module-axisartist
